src/dune/build_system.ml

```diff
@@ -1437,20 +1505,32 @@ end = struct
         in
         let chdirs = Action.chdirs action in
         Path.Set.iter chdirs ~f:Fs.(mkdir_p_or_check_exists ~loc);
-        let+ () =
+        let+ dynamic_deps_stages =
           with_locks locks ~f:(fun () ->
-            Fiber.map (Action_exec.exec action action_ctx) ~f:(fun () ->
-              Option.iter sandboxed ~f:(fun sandboxed ->
-                List.iter targets_as_list ~f:(fun target ->
-                  rename_optional_file ~src:(sandboxed target) ~dst:target))))
+            Fiber.map
+              (execute_action_until_all_deps_ready ~context ~env ~targets
+                ~rule_loc action) ~f:(fun used_deps ->
+                Option.iter sandboxed ~f:(fun sandboxed ->
```

Comment:
I think readability can be improved here by let-binding `copy_targets_out_of_sandbox` as you did in another PR (which involves rebasing, I guess)

Reply:
This is now included after merging this branch with master.